circuit-types/src/keychain.rs

```diff
@@ -129,105 +140,63 @@ impl From<SecretIdentificationKey> for Scalar {
     linkable,
     secret_share
 )]
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub struct NonNativeKey<const KEY_WORDS: usize> {
-    /// The `Scalar` words used to represent the key
-    ///
-    /// Because the key is a point on a non-native curve, its representation requires
-    /// a bigint like approach
-    pub key_words: [Scalar; KEY_WORDS],
-}
-
-impl<const KEY_WORDS: usize> Default for NonNativeKey<KEY_WORDS> {
-    fn default() -> Self {
-        Self {
-            key_words: [Scalar::zero(); KEY_WORDS],
-        }
-    }
+#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+pub struct PublicSigningKey {
+    pub x: [Scalar; FELT_WORDS],
+    pub y: [Scalar; FELT_WORDS],
 }
 
-impl<const KEY_WORDS: usize> NonNativeKey<KEY_WORDS> {
-    /// Split a biguint into scalar words in little endian order
-    fn split_biguint_into_words(mut val: BigUint) -> [Scalar; KEY_WORDS] {
-        let scalar_mod = get_scalar_field_modulus();
-        let mut res = Vec::with_capacity(KEY_WORDS);
-        for _ in 0..KEY_WORDS {
-            let word = Scalar::from(&val % &scalar_mod);
-            val /= &scalar_mod;
-            res.push(word);
-        }
+impl PublicSigningKey {
+    /// Split a biguint into 128-bit scalar words
+    fn split_biguint_into_words(val: &BigUint) -> [Scalar; FELT_WORDS] {
+        let low_mask = val & &*LOW_BIT_MASK;
+        let low: u128 = low_mask.try_into().unwrap();
+        let high: u128 = (val >> 128u32).try_into().unwrap();
 
-        res.try_into().unwrap()
+        [Scalar::from(low), Scalar::from(high)]
     }
 
-    /// Re-collect the key words into a biguint
-    fn combine_words_into_biguint(&self) -> BigUint {
-        let scalar_mod = get_scalar_field_modulus();
-        self.key_words
-            .iter()
-            .rev()
-            .fold(BigUint::from(0u8), |acc, word| {
-                acc * &scalar_mod + word.to_biguint()
-            })
-    }
-}
+    /// Combine 128-bit scalar words into a biguint
+    fn combine_words_into_biguint(words: &[Scalar; FELT_WORDS]) -> BigUint {
+        let low_biguint = words[0].to_biguint();
+        let high_biguint = words[1].to_biguint();
 
-impl<const KEY_WORDS: usize> From<&BigUint> for NonNativeKey<KEY_WORDS> {
-    fn from(val: &BigUint) -> Self {
-        Self {
-            key_words: Self::split_biguint_into_words(val.clone()),
-        }
+        (high_biguint << 128) + low_biguint
     }
 }
 
-impl<const KEY_WORDS: usize> From<&NonNativeKey<KEY_WORDS>> for BigUint {
-    fn from(value: &NonNativeKey<KEY_WORDS>) -> Self {
-        value.combine_words_into_biguint()
-    }
-}
+impl From<StarkPoint> for PublicSigningKey {
+    fn from(value: StarkPoint) -> Self {
+        let Affine { x, y, infinity } = value.to_affine();
+        if infinity {
+            Self {
+                x: [Scalar::zero(); 2],
+                y: [Scalar::zero(); 2],
+            }
```

Comment:
I believe the IETF spec disallows the additive identity as the public key? If so, we should `assert` here

Reply:
oh yes great call